msim/src/sim/task.rs

```diff
@@ -65,6 +67,49 @@ pub(crate) struct NodeInfo {
     span: Span,
 }
 
+#[derive(Debug)]
+struct PanicWrapper {
+    // how long should the node stay down.
+    restart_after: Duration,
+}
+
+struct PanicHookGuard(Option<Box<dyn Fn(&std::panic::PanicInfo<'_>) + Sync + Send + 'static>>);
+
+impl PanicHookGuard {
+    fn new() -> Self {
+        Self(Some(std::panic::take_hook()))
+    }
+
+    fn call_hook(&self, info: &std::panic::PanicInfo<'_>) {
+        self.0.as_ref().unwrap()(info);
+    }
+}
+
+impl Drop for PanicHookGuard {
+    fn drop(&mut self) {
+        std::panic::set_hook(self.0.take().unwrap());
+    }
+}
+
+/// Kill the current node by panicking with a special type that tells the executor to kill the
+/// current node instead of terminating the test.
+pub fn kill_current_node(restart_after: Option<Duration>) {
+    let handle = runtime::Handle::current();
+    let cur_node_id = context::current_node();
+
+    let restart_after = restart_after.unwrap_or_else(|| {
+        Duration::from_millis(handle.rand.with(|rng| rng.gen_range(1000..3000)))
```

Comment:
Is the random time just to make the network seem more organic in its recovery?

Reply:
pretty much, yes.